lib/ble.toit

```diff
@@ -171,339 +116,389 @@ class RemoteDevice:
   /**
   The UUID of the service.
-
-  For 16 and 32 bit UUIDs, form the BLE variant with the top-level uuid function.
   */
-  uuid/uuid_pkg.Uuid
+  uuid/BleUUID
+  peripheral_manager/PeripheralManager
 
-  server_configuration_/ServerConfiguration
-  resource_/ByteArray?
+  constructor .peripheral_manager .uuid:
+    resource := ble_add_service_ peripheral_manager.resource_ uuid.encode_for_platform_
+    super peripheral_manager.adapter.resource_group_ resource --auto_release
 
-  characteristics_/Map ::= {:}
+  /**
+  Adds a characteristic to this service with the given paraemters.
+  $uuid is the uuid of the characteristic
+  $properties is one of the CHARACTERISTIC_PROPERTY_* values
+  $permissions is one of the CHARACTERISTIC_PERMISSIONS_* values
+  if $value is specified, it will be used as an initial value for the characteristic.
 
-  constructor .server_configuration_ .uuid:
-    resource_ = ble_add_server_service_ server_configuration_.resource_group_ uuid.to_byte_array
+  NOTE: Only has effect if the corresponding service is not yet deployed
+  */
+  add_characteristic
+      uuid/BleUUID
+      --properties/int
+      --permissions/int
+      --value/ByteArray=#[] -> LocalCharacteristic:
+    return LocalCharacteristic this uuid properties permissions value
 
-  add_read_only_characteristic uuid --value=#[]-> ReadOnlyCharacteristic:
-    char := ReadOnlyCharacteristic this uuid value
-    characteristics_[uuid] = char
-    return char
+  /**
+  Adds a read only characteristic with the given $uuid and $value. See $add_characteristic.
+  */
+  add_read_only_characteristic uuid/BleUUID --value/ByteArray -> LocalCharacteristic:
+    return add_characteristic
+        uuid
+        --properties=CHARACTERISTIC_PROPERTY_READ
+        --permissions=CHARACTERISTIC_PERMISSION_READ
+        --value=value
 
-  add_write_only_characteristic uuid --require_response/bool=true -> WriteOnlyCharacteristic:
-    char := WriteOnlyCharacteristic this uuid require_response
-    characteristics_[uuid] = char
-    return char
+  /**
+  Adds a write only characteristic with the given $uuid that can $require_responsew for each write.
+  See $add_characteristic.
+  */
+  add_write_only_characteristic uuid/BleUUID requires_response/bool=false -> LocalCharacteristic:
+    return add_characteristic
+        uuid
+        --properties=requires_response?CHARACTERISTIC_PROPERTY_WRITE:CHARACTERISTIC_PROPERTY_WRITE_WITHOUT_RESPONSE
+        --permissions=CHARACTERISTIC_PERMISSION_WRITE
 
-  add_read_write_characteristic uuid --value=#[]-> ReadWriteCharacteristic:
-    char := ReadWriteCharacteristic this uuid value
-    characteristics_[uuid] = char
-    return char
+  /**
+  Adds a notification characteristic with the given $uuid that can be an indication if $indication is true,
+  otherwise it will be a notification. See $add_characteristic.
+  */
+  add_notification_characteristic uuid/BleUUID indication/bool=false -> LocalCharacteristic:
```

Comment:
If chosen to use the `--read_only` and `write_only` versions from above, might as well make two more:
`--for_notification` and `--for_indication`.  (Not 100% happy with the naming):
```
add_characteristic --for_notification ...
add_characteristic --for_indication ...
```